File: 6. Binary Search/4. Median of Two Sorted Arrays/Solution.py
from typing import List

class Solution:
    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:
        if len(nums2) < len(nums1):
            nums1, nums2 = nums2, nums1

        m = len(nums1)
        n = len(nums2)
        half_total = (m + n) // 2

        left = 0
        right = len(nums1) - 1

        while True:
            L1_index = left + (right - left) // 2
            L2_index = half_total - (L1_index + 1) - 1

            L1 = float('-inf') if L1_index < 0 else nums1[L1_index]
            R1 = float('inf') if L1_index + 1 > m - 1 else nums1[L1_index + 1]
            L2 = float('-inf') if L2_index < 0 else nums2[L2_index] 
            R2 = float('inf') if L2_index + 1 > n - 1 else nums2[L2_index + 1]

            if L1 > R2:
                right = L1_index - 1
            elif L2 > R1:
                left = L1_index + 1
            else:
                if (m + n) % 2 == 0:
                    return (max(L1, L2) + min(R1, R2)) / 2
                else:
                    return min(R1, R2)

File: 6. Binary Search/4. Median of Two Sorted Arrays/Solution.cpp
#include <vector>
#include <algorithm>
#include <cmath>

class Solution {
public:
    double findMedianSortedArrays(std::vector<int>& nums1, std::vector<int>& nums2) {
        if (nums2.size() < nums1.size()) std::swap(nums1, nums2);
        int m = nums1.size();
        int n = nums2.size();
        int halfSize = (m + n) / 2;

        int left = 0;
        int right = m - 1;

        while (true)
        {
            int leftIndex1 = (right + left < 0) ? -1 : (right + left) / 2;
            int leftIndex2 = halfSize - (leftIndex1 + 1) - 1;

            double left1 = (leftIndex1 < 0) ? -INFINITY : nums1[leftIndex1];
            double right1 = (leftIndex1 + 1 > m - 1) ? INFINITY : nums1[leftIndex1 + 1];
            double left2 = (leftIndex2 < 0) ? -INFINITY : nums2[leftIndex2];
            double right2 = (leftIndex2 + 1 > n - 1) ? INFINITY : nums2[leftIndex2 + 1];

            if (left1 > right2)
            {
                right = leftIndex1 - 1;
            }
            else if (left2 > right1)
            {
                left = leftIndex1 + 1;
            }
            else
            {
                if ((m + n) % 2 == 0)
                {
                    return (std::max(left1, left2) + std::min(right1, right2)) / 2.0;
                }
                else
                {
                    return std::min(right1, right2);
                }
            } 
        }
    }
};